src/NUnitFramework/framework/Internal/TestNameGenerator.cs

```diff
@@ -263,23 +269,40 @@ public override void AppendTextTo(StringBuilder sb, MethodInfo method, object?[]
             }
         }
 
-        private sealed class ClassNameFragment : NameFragment
+        private sealed class ClassNameFragment : NameFromMethodInfoFragment
         {
-            public override void AppendTextTo(StringBuilder sb, MethodInfo method, object?[]? args)
+            protected override void AppendTextTo(StringBuilder sb, MethodInfo method, object?[]? args)
             {
                 sb.Append(method.DeclaringType!.Name);
             }
         }
 
-        private sealed class ClassFullNameFragment : NameFragment
+        private sealed class ClassFullNameFragment : NameFromMethodInfoFragment
         {
-            public override void AppendTextTo(StringBuilder sb, MethodInfo method, object?[]? args)
+            protected override void AppendTextTo(StringBuilder sb, MethodInfo method, object?[]? args)
             {
                 sb.Append(method.DeclaringType!.FullName);
             }
         }
 
-        private sealed class ArgListFragment : NameFragment
+        // TODO: Don't think this quite works yet.
+        private sealed class DerivedClassNameFragment : NameFragment
+        {
+            public override void AppendTextTo(StringBuilder sb, TestMethod testMethod, object?[]? args)
+            {
+                sb.Append(testMethod.TypeInfo!.Name);
```

Comment:
I think the problem is that TypeInfo is set to be the methodInfo.TypeInfo.  Need to find another way of accessing the correct data.